### DISCUSSION TITLE
Reusing existing ontologies

### DISCUSSION THREAD

**Discussion by ngeiswei** [singnet/ai-dsl#18] [General]
## Overview

This discussion is about exploring the possibility of reusing existing ontologies, their data and conventions, not necessarily their
languages (thus possibility requiring porting their content to our to-be-determined language).

## Methodology

The ontologies in the section below have been primarily obtained from the follow 2 lists

1. https://protegewiki.stanford.edu/wiki/Protege_Ontology_Library
2. https://info.sice.indiana.edu/~dingying/Teaching/S604/OntologyList.html

plus extra search on the Internet.  I should mentioned that over half of the links in these lists were broken.  In some cases I have tried to search up-to-date links, and in some other cases (when the description didn't look promising) I passed.  I tried to be thorough but it's possible I have missed good candidates.

## Ontologies

The table below contain ontologies obtained from the methodology above, keeping the ones with the slightest chance of being at least
partially reusable.  I tried to sort them by relevance and up-to-date-ness.

| Name | URL | Description | Relevance (0-3) | Up-to-date (0-1) |
| ---- | --- | ----------- | --------------- | ---------------- |
| Software Ontology (SWO) | http://theswo.sourceforge.net/ | Describe software tools, their types, tasks, versions, provenance and data associated | 3 | 1 |
| Computer Science Ontology (CSO) | https://cso.kmi.open.ac.uk/home | Taxonomy generated from articles of Computer Science research topics | 2 | 1 |
| Suggested Upper Merged Ontology (SUMO) | http://www.ontologyportal.org/ | Very broad "largest formal public ontology in existence" | 2 | 1 |
| Information Artifact Ontology (IAO) | https://github.com/information-artifact-ontology/IAO/ | Ontology of information entities | TBD | 1 |
| Basic Formal Ontology (BFO) | https://basic-formal-ontology.org/ | Upper-level ontology for information retrieval, analysis and integration in scientific domains | TBD | 1 |
| Web Service Modeling Ontology (WSMO) | https://www.w3.org/Submission/WSMO/ | Framework for describing aspects of Web services | 1 | 0 |
| The Event Ontology | http://motools.sourceforge.net/event/event.html | Centered around the notion of spacetime event | 1 | 0 |
| Core Ontology for Multimedia Annotation (COMM) | https://www.researchgate.net/publication/226736019_COMM_A_Core_Ontology_for_Multimedia_Annotation | Paper on an ontology for multimedia annotation | TBD | 0 |
| WordNet | https://wordnet.princeton.edu/ | Large lexical database of English | TBD | 1 |

## More

We might want to go through the following list as well, even though it is centered around bio-informatics.

* https://bioportal.bioontology.org/ontologies

**Comment by ngeiswei** [comment]
I had a deeper look at SUMO.  There are some potentially reusable data about computer technology in the following files

https://github.com/ontologyportal/sumo/blob/master/CCTrep.kif
https://github.com/ontologyportal/sumo/blob/master/ComputerInput.kif
https://github.com/ontologyportal/sumo/blob/master/ComputingBrands.kif
https://github.com/ontologyportal/sumo/blob/master/Media.kif
https://github.com/ontologyportal/sumo/blob/master/QoSontology.kif
https://github.com/ontologyportal/sumo/blob/master/engineering.kif

Here are examples of relationships you can find

```
(subclass Sensor ComputerHardware)

(subclass UniformResourceLocator InternetAddress)

(=>
	(instance ?URL UniformResourceLocator)
	(exists (?Address)
		(and
			(instance ?Address IPAddress)
			(represents ?URL ?Address))))

(subclass FlashDrive SolidStateDataStorageDevice)

(instance runningOn BinaryPredicate)
(domain runningOn 1 ComputerProgram)
(domain runningOn 2 Computer)

(subclass LinearEquation Equation)
```

This is a tiny snippet of all that is described. Yet it's unclear we may find anything reusable but we may want to consider it.

**Comment by kabirkbr** [comment]
I suggest to conceptualize the issue of choosing ontologies to reuse a bit.

## Purpose of AI-DSL ontology

The overarching purpose of AI-DSL is to facilitate search, matching and communication among decentralized AI agents on SingularityNET platform. However, for the purposes of DSL design, we can think in more general terms -- about communication between any AI agents / algorithms.  

For communication, ontology is needed in order to compare instances (of data / types) to each other and to resolve:
1. if these instances are of the same class;
1. can we prove equivalence relation between these instances (say input data / type == output data type);
1. can we convert one instance to another without any loss of expressiveness / functionality which would allow the previous point;
2. can we have some sort of distance measure between complex instances of (data/types) so that the probability of match of input to output could be established, even if the formal equivalence cannot;

The purpose description given here does not try to be complete or nearly so, but is just an indication. However we need to have good indication what we are going to use this ontology for before going further, as talked quite a few times before. Having done that (which is not yet complete), we can try choosing the following;

## Ontology language

It seems to make sense to think about ontology languages before ontologies, or at least in parallel. Note, the relation between ontology languages and ontologies are of many-to-many type, i.e. the same ontology can be expressed in more than one ontology language and ontology language of course can express many ontologies. The point is that choosing a good ontology to reuse does not solve the question which language we choose. But language is important for even incompletely described AI-DSL purposes above.

One specific requirement for ontology language which allows to use it in uncertain environment is the ability to convert it to other formats / languages. If we find language like this, then we may not need to think too much about formal properties too much, since we will always have flexibility to switch to another language.

[2](https://link.springer.com/chapter/10.1007/978-3-642-24511-4_10) provides a comparison between most popular (at least as they claim) ontology languages:

![image](https://user-images.githubusercontent.com/1760288/116589172-a8bbf200-a91c-11eb-91a8-d0c437143a11.png)

## Reusing existing ontologies

There are three types of ontologies: uper-level (something that I called ontology schema in the past, but this is probably not a correct name for it), mid-level and lower-level. The working ontology most probably is lower-level, but we need to choose all levels in order to have the lowest levels, in my opinion. The table of ontologies given in the Nil's post does not seem to discriminated entries by these levels. 

<img src="https://user-images.githubusercontent.com/1760288/116587383-b07a9700-a91a-11eb-997c-9ad8c8ce7052.png" width="50%">

From this [1](https://www.researchgate.net/publication/221292703_Survey_on_Ontology_Languages):
1. Upper ontologies are high-level, domain-independent ontologies that provide common knowledge bases from which more domain-specific ontologies may be derived. Standard upper ontologies are also referred to as foundational or universal ontologies.
2. Mid-level ontologies are less abstract and make assertions that span multiple domain ontologies. These ontologies may provide more concrete representations of abstract concepts found in the upper ontology. There is no clear demarcation point between upper and mid-level. Mid-level ontologies also encompass the set of ontologies that represent commonly used concepts, such as Time and Location. These commonly used ontologies are sometimes referred to as utility ontologies.
3. Doman ontologies specify concepts particular to a domain of interest and represent those concepts and their relationships from a domain specific perspective. Domain ontologies may be composed by importing mid-level ontologies. They may also extend concepts defined in mid-level or upper ontologies.

It seems to me that it makes sense to choose upper level ontology first and then go for searching lower level ontologies, because there is a good chance that we will not find anything suitable there. In this case we will need to develop our own mid-level and domain ontology, for which we will need an upper ontology.

### Upper level ontologies

Wikipedia gives a full list of avialable upper level ontologies https://en.wikipedia.org/wiki/Upper_ontology#Available_upper_ontologies

My hunch, based on limited reading is to choose SUO-KIF as ontology language and then SUMO as upper-level ontology and then see if we can develop something for AI-DSL based on examples. An important criteria is that SUMO / SUO-KIF seems to be the portability to other languages. This is not hugely informed by other criteria, therefore more thinking discussion maybe welcome.

**Comment by kabirkbr** [comment]
We may want to have a look at this https://adimen.si.ehu.es/web/adimenSUMO (can be for the second iteration)

![image](https://user-images.githubusercontent.com/1760288/120501223-bbb15e80-c3c1-11eb-8615-f84d637f392e.png)

**Comment by ngeiswei** [comment]
Another potential good source of ontology is [Media Type](https://en.wikipedia.org/wiki/Media_type).  Examples include

[image/gif](https://www.iana.org/assignments/media-types/image/gif)
[audio/vorbis](https://www.iana.org/assignments/media-types/audio/vorbis)
[video/mpeg](https://www.iana.org/assignments/media-types/video/mpeg)

and so many more.
